upbc/protoc-gen-upb.cc

```diff
@@ -980,11 +1130,74 @@ void WriteMessage(const protobuf::Descriptor* message, Output& output,
   output("};\n\n");
 }
 
-void WriteMessages(const protobuf::FileDescriptor* file, Output& output,
+void WriteExtension(const protobuf::FieldDescriptor* ext, Output& output) {
+  output("const upb_msglayout_ext $0 = {\n  ", ExtensionLayout(ext));
+  WriteField(ext, "0", "0", 0, output);
+  output(",\n");
+  output("  &$0,\n", MessageInit(ext->containing_type()));
+  if (ext->message_type()) {
+    output("  {.submsg = &$0},\n", MessageInit(ext->message_type()));
+  } else {
+    output("  {.submsg = NULL},\n");
+  }
+  output("\n};\n");
+}
+
+int WriteMessages(const protobuf::FileDescriptor* file, Output& output,
                    bool fasttable_enabled) {
-  for (auto* message : SortedMessages(file)) {
+  std::vector<const protobuf::Descriptor*> file_messages =
+      SortedMessages(file);
+
+  if (file_messages.empty()) return 0;
+
+  for (auto message : file_messages) {
     WriteMessage(message, output, fasttable_enabled);
   }
+
+  output("static const upb_msglayout *$0[$1] = {\n", kMessagesInit,
+         file_messages.size());
+  for (auto message : file_messages) {
+    output("  &$0,\n", MessageInit(message));
+  }
+  output("};\n");
+  output("\n");
+  return file_messages.size();
+}
+
+int WriteExtensions(const protobuf::FileDescriptor* file, Output& output) {
+  auto exts = SortedExtensions(file);
+  absl::flat_hash_set<const protobuf::Descriptor*> forward_decls;
+
+  if (exts.empty()) return 0;
+
+  for (auto ext : exts) {
+    forward_decls.insert(ext->containing_type());
+    if (ext->message_type()) {
+      forward_decls.insert(ext->message_type());
+    }
+  }
+
+  for (auto ext : exts) {
+    WriteExtension(ext, output);
+  }
+
+  for (auto decl : forward_decls) {
```

Comment:
this will cause non-determinism in the output.  I think you should actually do this in a one pass thing.

auto exts = SortedExtensions(file);
if (exts.empty()) return 0;

absl::flat_hash_set<const protobuf::Descriptor*> seen_forward_decls = { nullptr };
auto output_forward_decl = [&](const auto* decl) {
  if (seen_forward_decls.insert(decl).second) {
    output("extern const upb_msglayout $0;\n", MessageInit(decl));
  }
};

for (auto ext : exts) {
  output_forward_decl(ext->containing_type());
  output_forward_decl(ext->message_type());
}

Reply:
Great catch. Your pattern is very nice and efficient, but for consistency with other forward declarations in this file I made it two-pass so that the types could be sorted by name. I did use the lambda idea though. :)